### PR TITLE
Implement replace(), split(), and substring() String Functions

### PR DESCRIPTION
**Pull request by xiazcy**
Final set of string function implementations from the [proposal](https://github.com/apache/tinkerpop/blob/master/docs/src/dev/future/proposal-3-remove-closures.asciidoc#string-manipulation-functions-in-tinkerpop). Not considering scopes at this point.

### CODE REVIEW

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SplitStepTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+public class SplitStepTest extends StepTest {
+
+    @Override
+    protected List<Traversal> getTraversals() {
+        return Collections.singletonList(__.split("a"));
+    }
+
+    @Test
+    public void testReturnTypes() {
+        assertArrayEquals(new String[]{"h", "llo"}, __.__("hello").split("e").next().toArray());
```
Comment (vkagamlyk) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SplitStepTest.java:43]: add test for `__.__("abc|abc").split("c|a")`

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    protected List<Traversal> getTraversals() {
+        return Arrays.asList(
+                __.substring(1L),
+                __.substring(1L, 2L));
+    }
+
+    @Test
+    public void testReturnTypes() {
+        assertEquals("ello wor", __.__("hello world").substring(1L, 8L).next());
```
Comment (vkagamlyk) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java:44]: add test for `__.__("hello world").substring(0L, 2147483648L).next()`

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                Math.toIntExact(Math.min(this.length + newStart, strItem.length())));
+    }
+
+    @Override
+    public Set<TraverserRequirement> getRequirements() {
+        return Collections.singleton(TraverserRequirement.OBJECT);
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java:90]: may be better to start using recommended java way for calculation hashcode? 
something like `result = 31 * result + ...`

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ReplaceStepTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+public class ReplaceStepTest extends StepTest {
+
+    @Override
+    protected List<Traversal> getTraversals() {
+        return Collections.singletonList(__.replace("a", "b"));
+    }
+
+    @Test
+    public void testReturnTypes() {
+        assertEquals("halo", __.__("hello").replace("el", "a").next());
```
Comment (vkagamlyk) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ReplaceStepTest.java:43]: good to have tests for
`__.__("hello").replace("", "a")`
`__.__("hello").replace(null, "a")`
`__.__("hello").replace("el", null)`

**Comment by kenhuuu** [comment]
You should add tests for null and empty string as input parameters (where applicable) to all the functions.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java (hunk 2 above)
Comment (xiazcy) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStepTest.java]: Changed to take `int` arguments so not applicable.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java (hunk 3 above)
Comment (xiazcy) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SubstringStep.java:89]: Updated

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ReplaceStepTest.java (hunk 4 above)
Comment (xiazcy) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/ReplaceStepTest.java:45]: Updated implementation here, following `StringUtils.replace()` logic, so `null` will be no-op to return the original string.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SplitStepTest.java (hunk 1 above)
Comment (xiazcy) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/SplitStepTest.java:45]: Added, and updated implementation to parse it properly as string (instead of regex)

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java [hunk 5]
```diff
@@ -1437,6 +1440,66 @@ public default GraphTraversal<S, String> concat(final String... concatStrings) {
...
+    /**
+     * Returns a list of strings created by splitting the incoming string traverser around the matches of the given delimiter.
+     * Null values are not processed and remain as null when returned. If the incoming traverser is a non-String
+     * value then an {@code IllegalArgumentException} will be thrown.
+     *
+     * @return the traversal with an appended {@link SplitStep}.
+     * @see <a href="http://tinkerpop.apache.org/docs/${project.version}/reference/#split-step" target="_blank">Reference Documentation - Split Step</a>
+     * @since 3.7.1
+     */
+    public default GraphTraversal<S, List<String>> split(final String splitChar) {
```
[the remark this answers is not in the document]
Comment (xiazcy) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java]: Updated to `separator`

**Comment by vkagamlyk** [comment]
thanks for your work!
VOTE+1

**Comment by Cole-Greer** [comment]
Thanks Yang, LGTM.

VOTE +1

**Comment by kenhuuu** [comment]
VOTE +1
